### ISSUE TITLE
JetBrains IntelliJ IDEA and other IDEs disappearing from alttab

### ISSUE THREAD

**Issue by dvcrn** [lwouis/alt-tab-macos#1249]
**Describe the bug**

Continuation from https://github.com/lwouis/alt-tab-macos/issues/824

I have repeatedly the issue that jetbrains stuff just vanishes from the alt-tab after longer working sessions. Not doing any fullscreen settings here and keep it in a normal window all the time. 

~Only after restarting IDEA, it shows back up. Restarting alt-tab does not help.~ Restarting either alt-tab or IDEA fixes this

IDEA and other apps are installed through jetbrains toolbox

**Screenshots / video**

<img width="686" alt="Screen Shot 2021-11-24 at 11 09 23" src="https://user-images.githubusercontent.com/688326/143158693-e9504a09-7822-4d19-979d-abf141381601.png">

<img width="686" alt="Screen Shot 2021-11-24 at 11 09 24" src="https://user-images.githubusercontent.com/688326/143158671-eb7b608d-d032-4141-a8e1-ac0537fae11c.png">

**Steps to reproduce the bug**

I'm not actually sure. I tested this with CLion, IDEA Ultimate and Goland, and they all disappear after a longer working session (a couple hours)

**Comment by lwouis** [comment]
> Restarting alt-tab does not help.

This is interesting/unexpected. usually issues with windows missing from AltTab are because AltTab may have missed an event or there was a transition. Here it doesn't show after a restart.

It would seem to indicate that the windows not shown on AltTab may be considered "not real windows". Could you please show screenshots/videos of exactly which windows are not showing but you think should show. The best would be a video where you show the app windows, AltTab thumbnails, and narrate what you think is wrong

**Comment by Dimezis** [comment]
I have the same problem both with IntelliJ IDEA and Android Studio (basically the same thing).
Except restarting alt-tab always helps in my case.
Basically, if the IDE was started after alt-tab, it's never shown. If it's started before alt-tab, it's being detected just fine.

**Comment by dvcrn** [comment]
Thanks for chiming in! I just tried it again and yu are right, restarting alt-tab does help. Sorry for misreporting this. 

JetBrains stuff is getting sorted at the end of the alt-tab switcher so I didn't notice it was in the list. So yeah I confirm what @Dimezis said, restarting alt-tab does fix this

**Comment by lwouis** [comment]
Can one of you please share a video so i can see which windows are having this issue?

I need more data to investigate. Video, screenshot, what's specific about these windows, when does it happen, etc.

A first step would be to share a video 👍

**Comment by Dimezis** [comment]
https://drive.google.com/file/d/17aqCcFlzpOMae7xTCkvppbKLSvtCCldW/view?usp=sharing
Here's a short video of this bug. 
I _think_ it might be related to that "Android Studio Preview needs to be updated" popup.
Sometimes, when this popup doesn't appear before the AS is being opened, AltTab detects it fine.

**Comment by dvcrn** [comment]
@Dimezis I checked out your video, but I never have the "needs to be updated" popup because I manage my stuff through the toolbox

I tried some actions here that cause a popup to appear (like invalidate caches), but it's still appearing fine for me in Alt-Tab. Couldn't figure out yet what's causing it to disappear. Could you try `Invalidate Caches` on your end after restarting alt-tab to see if that's also making it disappear?

**Comment by lwouis** [comment]
@Dimezis so you're saying the windows not showing in AltTab is the following:

* It happens only when the "Android Studio Preview needs to be updated" popup showed at launch
* It doesn't show any window then

There are no other scenarios? It seems like it's very specific. Probably different from what @dvcrn is facing.

Could you please share the BundleIdentifier of your Android Studio app? (see [how to get it on that thread](https://superuser.com/a/1004714/1163149))

**Comment by Dimezis** [comment]
> It happens only when the "Android Studio Preview needs to be updated" popup showed at launch

Hard to say, I mean this popup appears only once in a while. But when it doesn't appear on subsequent launches, the AS is being detected fine.
Bundle ID  - `com.google.android.studio-EAP%`

**Comment by jankatins** [comment]
I can also confirm that the jet brains editor (intelij in my case) comes back after a start of alt-tab. Anything you need to debug this further?

**Comment by mkocztorz** [comment]
I have same issue. I restarted the entire system and even then problem persisted. However, when I came across this issue I restarted alt-tab and intelij windows are back.
alt-tab Version 6.29.0
macOS 11.6 (20G165) Big Sur

**Comment by teamhide** [comment]
Any update here?

**Comment by CodeMason** [comment]
Wanted to mention that I am also experiencing this issue.

M1 Max: 12.1
AltTab Version 6.29.0
IntelliJ IDEA 2021.3.1

**Comment by luyi404** [comment]
Same with GoLand :(
Maybe the JetBrains apps wanna hide from being detected?

M1 Pro Mbp2021 16inch,
macOS: 12.1
AltTab: 6.29.0
GoLand: 2021.3.2

**Comment by lwouis** [comment]
@luyi404 most likely it's due to Jetbrains product using a Java UI framework instead of native AppKit from Apple. This creates all sorts of problems for accessibility apps like AltTab as a window is often not coded as a NSWindow. Or Dialogs are codes as NSWindow. Etc

**Comment by luyi404** [comment]
@lwouis Thanks for the reply!

I was wondering if this could be solved by some hard coding? Like doing additional checks for JetBrains products or just apps with Java UI framework (I'm not familiar with client-side development for the macOS platform, so maybe what I'm saying isn't quite feasible) 😃

**Comment by CodeMason** [comment]
@lwouis @luyi404 i do not think this it this has to do with java. I have other full screen apps that do the same thing, primarily games (that are very obviously not java).

**Comment by dvcrn** [comment]
That's the same problem down at the core, no? Those likely also don't use AppKit as their UI framework

I had other issues with Java-based apps and accessibility tools in the past

**Comment by jameslkingsley** [comment]
Perhaps an option in AltTab to listen for when an app is launched, and then silently restart AltTab? Restarting always fixes it for me

**Comment by lwouis** [comment]
@jameslkingsley instead of restarting we could run whatever code is run at start. That would be a bit more reasonable 😅. That being said this approach wouldn't work. Windows come and go at any time, not just when an app is launched. We would need to run this on a schedule since we wouldn't have a particular event to trigger with. In addition, another problem is that on startup AltTab guesses the last-order of windows based on their z-index. It's a guess and is not 100% accurate. Then once it's running it then maintains a correct list of windows and when they have been used last. That one is reliable but comes from observing all sorts of events, which the startup technique has no idea about.

**Comment by jesusfj710** [comment]
Any chance of getting this fixed? I am able to reproduce it with IntelliJ IDEA and WebStorm Idea. I can provide more information if you tell me what you need.

**Comment by weihao** [comment]
Same issue here.

Following

**Comment by vrnvorona** [comment]
@metacodes hope this fix works well and will be merged 🚀

**Comment by nt-jjrubio** [comment]
Same issue here.

Following

**Comment by lwouis** [comment]
Should be fixed in v6.32.0

**Comment by AlexZeitler** [comment]
Thanks!

I'm using 6.32.0 since yesterday and I didn't face the issue again (until the update it happened multiple times a day).

**Comment by perdian** [comment]
I can confirm that the latest version (for me that's 6.33.0) solved the issue for me as well. 

Thanks 👍

**Comment by dvcrn** [comment]
Oh nice! I'll give 6.33 a try then and see how it goes 👍

**Comment by AlexZeitler** [comment]
Looks like this is back with https://github.com/lwouis/alt-tab-macos/releases/tag/v6.35.0 (will try to find out which version exactly)

**Comment by lwouis** [comment]
@AlexZeitler please open a new ticket with steps to reproduce and a video or screenshots. There is hope to fix only if we have a way to reproduce it consistently.

Also note the discussions about deep architecture changes we would need to do to better support some use-cases: https://github.com/lwouis/alt-tab-macos/issues/447#issuecomment-1112805183 and https://github.com/lwouis/alt-tab-macos/pull/1484
